File: src/main/java/com/moneydance/modules/features/formula/reminder/RemindersListTableModel.java
package com.moneydance.modules.features.formula.reminder;

import com.infinitekind.moneydance.model.Reminder;
import com.infinitekind.util.CustomDateFormat;
import com.moneydance.modules.features.formula.MDApi;

import javax.swing.table.AbstractTableModel;
import java.util.ArrayList;
import java.util.List;

public class RemindersListTableModel extends AbstractTableModel {

    private final Reminder NEW_ENTRY = new Reminder(null);

    private List<Reminder> reminders = new ArrayList<>();

    private CustomDateFormat dateFormat;

    public RemindersListTableModel(MDApi api) {
        dateFormat = api.getShortDateFormatter();
    }

    public void setReminders(List<Reminder> reminders) {
        this.reminders = reminders;
        fireTableDataChanged();
    }

    Reminder getReminder(int index) {
        return reminders.get(index);
    }

    int insert() {
        int insertIndex = reminders.size();
        reminders.add(insertIndex, NEW_ENTRY);
        fireTableRowsInserted(insertIndex, insertIndex);
        return insertIndex;
    }

    void remove(int index) {
        Reminder removed = reminders.remove(index);
        MDApi.disableReminder(removed);

        fireTableRowsDeleted(index, index);
    }

    @Override
    public int getRowCount() {
        return reminders.size();
    }

    @Override
    public int getColumnCount() {
        return 2;
    }

    @Override
    public String getColumnName(int column) {
        switch (column) {
            case 1:
                return "Next payment";
            default:
                return "Name";
        }
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex) {
        Reminder reminder = reminders.get(rowIndex);
        switch (columnIndex) {
            case 1:
                return dateFormat.format(reminder.getNextOccurance(29991231));
            default:
                return reminder.getDescription();
        }
    }

    @Override
    public boolean isCellEditable(int rowIndex, int columnIndex) {
        return columnIndex == 0 && NEW_ENTRY.equals(reminders.get(rowIndex));
    }

    @Override
    public void setValueAt(Object aValue, int rowIndex, int columnIndex) {
        if (!isCellEditable(rowIndex, columnIndex)) {
            return;
        }

        Reminder reminder = (Reminder)aValue;
        MDApi.enableReminder(reminder);

        reminders.set(rowIndex, reminder);
        fireTableRowsUpdated(rowIndex, rowIndex);
    }
}

File: src/main/kotlin/com/moneydance/modules/features/formula/reminder/RemindersListTableModel.kt
package com.moneydance.modules.features.formula.reminder

import com.moneydance.modules.features.formula.MDApi.Companion.disableReminder
import com.moneydance.modules.features.formula.MDApi.Companion.enableReminder
import com.moneydance.modules.features.formula.MDApi
import javax.swing.table.AbstractTableModel
import com.infinitekind.moneydance.model.Reminder
import com.infinitekind.util.CustomDateFormat
import java.util.ArrayList

class RemindersListTableModel(api: MDApi) : AbstractTableModel() {
    private val NEW_ENTRY = Reminder(null)
    private var reminders: MutableList<Reminder> = ArrayList()
    private val dateFormat: CustomDateFormat

    init {
        dateFormat = api.shortDateFormatter
    }

    fun setReminders(reminders: List<Reminder>) {
        this.reminders = reminders.toMutableList()
        fireTableDataChanged()
    }

    fun getReminder(index: Int): Reminder {
        return reminders[index]
    }

    fun insert(): Int {
        val insertIndex = reminders.size
        reminders.add(insertIndex, NEW_ENTRY)
        fireTableRowsInserted(insertIndex, insertIndex)
        return insertIndex
    }

    fun remove(index: Int) {
        val removed = reminders.removeAt(index)
        disableReminder(removed)
        fireTableRowsDeleted(index, index)
    }

    override fun getRowCount(): Int {
        return reminders.size
    }

    override fun getColumnCount(): Int {
        return 2
    }

    override fun getColumnName(column: Int): String {
        return when (column) {
            1 -> "Next payment"
            else -> "Name"
        }
    }

    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any {
        val reminder = reminders[rowIndex]
        return when (columnIndex) {
            1 -> dateFormat.format(reminder.getNextOccurance(29991231))
            else -> reminder.description
        }
    }

    override fun isCellEditable(rowIndex: Int, columnIndex: Int): Boolean {
        return columnIndex == 0 && NEW_ENTRY == reminders[rowIndex]
    }

    override fun setValueAt(aValue: Any, rowIndex: Int, columnIndex: Int) {
        if (!isCellEditable(rowIndex, columnIndex)) {
            return
        }
        val reminder = aValue as Reminder
        enableReminder(reminder)
        reminders[rowIndex] = reminder
        fireTableRowsUpdated(rowIndex, rowIndex)
    }
}